WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs

```diff
@@ -0,0 +1,95 @@
+using System.Linq;
+using System.Collections.Immutable;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Affiliation.Extensions;
+using WalletWasabi.Bases;
+using WalletWasabi.Extensions;
+
+namespace WalletWasabi.Affiliation;
+
+public class AffiliateServerStatusUpdater : PeriodicRunner
+{
+	private static readonly TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(20);
+	private static readonly TimeSpan Interval = TimeSpan.FromMinutes(1);
+
+	public AffiliateServerStatusUpdater(IDictionary<string, AffiliateServerHttpApiClient> clients)
+		  : base(Interval)
+	{
+		Clients = clients;
+	}
+
+	private IDictionary<string, AffiliateServerHttpApiClient> Clients { get; }
+	private HashSet<string> RunningAffiliateServers { get; } = new();
+	private object RunningAffiliateServersLock { get; } = new();
+	
+	public ImmutableArray<string> GetRunningAffiliateServers()
+	{
+		lock (RunningAffiliateServersLock)
+		{
+			return RunningAffiliateServers.ToImmutableArray();
+		}
+	}
+
+	protected override async Task ActionAsync(CancellationToken cancellationToken)
+	{
+		await UpdateRunningAffiliateServersAsync(cancellationToken).ConfigureAwait(false);
+	}
+
+	private static async Task<bool> IsAffiliateServerRunningAsync(AffiliateServerHttpApiClient client, CancellationToken cancellationToken)
+	{
+		try
+		{
+			await client.GetStatusAsync(cancellationToken).ConfigureAwait(false);
+			return true;
+		}
+		catch (Exception exception)
+		{
+			Logging.Logger.LogError(exception);
+			return false;
+		}
+	}
+
+	private async Task UpdateRunningAffiliateServersAsync(string affiliationId, AffiliateServerHttpApiClient affiliateServerHttpApiClient, CancellationToken cancellationToken)
+	{
+		using var linkedCts = cancellationToken.CreateLinkedTokenSourceWithTimeout(AffiliateServerTimeout);
+		
+		if (await IsAffiliateServerRunningAsync(affiliateServerHttpApiClient, linkedCts.Token).ConfigureAwait(false))
+		{
+			lock (RunningAffiliateServersLock)
+			{
+				if (!RunningAffiliateServers.Add(affiliationId))
```

Comment:
Should be `if (RunningAffiliateServers.Add(affiliationId))`.

Reply:
yep. done.